Sprint 4

dmirain opened the pull request:
Я не стал проходить урок до конца, а сам написал всё, как смог. Потом прошёл по курсу и что-то поправил, но многое осталось, так как я вижу. У меня всё было разложено по папочкам, но в требованиях написано, что надо сложить в один файлик, так и сделал. Не стал делать игру повторяющейся, а сделал рандомный генератор условий.

Commit:
Первая реализация игры без статистики

Commit:
Победил шрифты

Commit:
Всё в одном файле :(

Commit:
Не показываю рамку, когда нет ответа

IBDesignable commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,255 @@
+
 final class MovieQuizViewController: UIViewController {
-    // MARK: - Lifecycle
+    private var movieQuiz = MovieQuizModel(
+        riddleGenerator: RiddleGenerator(
+            imdbGateway: IMDBGateway()
+        )
+    )
```
Давай договоримся до +- стандартного порядка функций переменных: сначала перечисляем свойства класса, затем функции. При этом также группируем на приватные и публичные блоки.  Функции перечисляют в таком порядке: `init`ы, `override`ы, блоки публичных и приватных функций, и далее блок c аннотацией `@IBAction`

IBDesignable commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,255 @@
+        if movieQuiz.gameIsEnded {
+            showEndGameAlert()
+            return
+        }
```
если нужно сделать проверку и преждевременно выйти, если проверка не выполнилась – используют оператор guard. Для него придется инвертировать условие, но выглядеть будет "swifty"

IBDesignable commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,255 @@
+        imageView.layer.borderWidth = 0
+        textLabel.text = movieQuiz.currentRiddle.text
+        noButton.isEnabled = true
+        yesButton.isEnabled = true
```
Этот блок, вероятно, будет повторяться много раз, имеет смысл вынести в отдельную функцию с аргументом

IBDesignable commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,255 @@
+    @IBOutlet private weak var imageView: UIImageView!
+    @IBOutlet private weak var textLabel: UILabel!
+    @IBOutlet private weak var noButton: UIButton!
+    @IBOutlet private weak var yesButton: UIButton!
```
ожидаю все свойства увидеть в начале класса

IBDesignable commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,255 @@
+    @IBAction private func yesButtonClicked() {
+        performResultButtonClick(with: .yes)
+    }
+}
```
очень хорошее моделирование через енам 👍🏻

IBDesignable commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,255 @@
+        currentRiddle = movieRiddles[0]
+    }
+    
+    mutating func checkAnswer(_ answer: Answer) -> Bool {
```
Со структурами не приятно mutating-операции производить. Думаю здесь стоит описать это классом (хранит стейт, меняет его, выполняет вычисления – все признаки менеджера)

Ну и вероятно это тебе поможет в следующих спринтах подстроиться под требования

IBDesignable commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,255 @@
+}
 
+enum RiddleSign: String {
+    case More = "больше"
+    case Less = "меньше"
```
в swift в верхнем camelCase пишут только типы: кейс енама таковым не является, давай поправим на нижний camelCase

IBDesignable replied:
Это надо будет поправить

IBDesignable commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,255 @@
- Ответ: НЕТ
- */
+struct IMDBGateway {
+    func movies() -> [IMDBMovie] {
```
Хорошая декомпозиция структуры данных

IBDesignable commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,255 @@
+    }
+    
+    var text: String {
+        return "Рейтинг этого фильма \(riddleSign.rawValue) чем \(riddleValue)?"
```
Попробуй избавиться еще от лишнего нуля для целого riddleValue

IBDesignable commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,255 @@
+    }
+    
+    mutating func checkAnswer(_ answer: Answer) -> Bool {
+        var result = false
+        if currentRiddle.correctAnswer == answer {
+            correctAnswers += 1
+            result = true
+        }
```
В другом порядке будет немного читаемее:
```suggestion
        var isResultCorrect = currentRiddle.correctAnswer == answer
        if isResultCorrect {
            correctAnswers += 1
        }
```

IBDesignable commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,255 @@
+        yesButton.titleLabel?.font = mediumFont
+    }
+    
+    private func updateView() {
```
Хотелось бы здесь видеть более чистую функцию: получать модель состояния игры в качестве аргумента. 

Иначе название не совсем отражает суть происходящего

IBDesignable replied:
Это надо будет поправить

Commit:
Поправил по ревью:
- Переделал на state machine с подобием генератора
- Упорядочил методы и свойства в классах
- использую guard там, где это нужно
- вынес в методы повторяющиеся изменения свойств view
- Исправил Capitalize case.
- Вывожу оценку в загадке без 0 после запятой

Commit:
Добавил перемешивание фильмов

Commit:
Исправил видимость атрибутов

IBDesignable commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -26,61 +34,75 @@ final class MovieQuizViewController: UIViewController {
             self.movieQuiz.reset()
-            self.updateView()
+            let gameState = self.movieQuiz.nextGameState()
+            self.updateViewState(to: gameState)
```
все содержимое можно вынести в функцию restart/resetGame – ну и в целом придерживаться такого правила для кложур удобно: меньше работы с сильным self

IBDesignable commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -26,61 +34,75 @@ final class MovieQuizViewController: UIViewController {
+    }
+    
+    private func updateButtonsState(to state: GameState) {
+        switch state {
+        case .nextRiddle:
+            noButton.isEnabled = true
+            yesButton.isEnabled = true
+        case .negativeAnswer, .positiveAnswer, .gameEnded:
+            noButton.isEnabled = false
+            yesButton.isEnabled = false
+        }
```
Для GameState можно дописать разширение: 

```swift 
var canAnswer: Bool { 
switch self { 
case .nextRiddle: return true 
... 
}
}
```
и тогда функция упростится в две строки для каждой кнопки: `noButton.isEnabled = state.canAnswer`

IBDesignable commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -92,49 +114,52 @@ final class MovieQuizViewController: UIViewController {
+    private var movieRiddles: [MovieRiddle] = []
+    private var correctAnswers: Int = 0
+    private var gameIsEnded: Bool = false
+    private var currentRiddleIndex: Int = -1
```
-1 хорошо - решает задачу, но можно круче: 

не лучшая практика заводить невалидный стейт, ведь индекс по бизнес-логике из множества беззнаковых целых.  Для моделирования неопределенного состояния в swift есть опционалы, можно было бы завести optional int. 

Если пока сложно – можно оставить как есть)

IBDesignable commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -92,49 +114,52 @@ final class MovieQuizViewController: UIViewController {
+    case nextRiddle(MovieRiddle, Int, Int)
+    case positiveAnswer
+    case negativeAnswer
+    case gameEnded(Int, Int)
```
к ассоциированным типам хорошо еще добавлять метки: gameEnded(result: Int, count: Int) - чтобы понимать что в эти поля пихать)

IBDesignable commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -250,6 +275,20 @@ struct IMDBGateway {
+
+extension Double {
+    func formatForRiddle() -> String {
+        let numberFormatter = NumberFormatter()
+        numberFormatter.numberStyle = .decimal
+        numberFormatter.minimumIntegerDigits = 1
+        numberFormatter.minimumFractionDigits = 0
+        numberFormatter.maximumFractionDigits = 1
```
С любыми форматтерами нужно быть аккуратными – они могут быть достаточно затратными, если часто вызывать эту функцию, чтобы получить число можно накопить неоптимальную работу приложения.
Общая рекомендация к любым форматтерам: дат, чисел, форматов – создать и настроить один раз под одну повторяющуюся задачу. 

Посмотреть как это можно делать, можно в Helpers у расширения к DateFormatter

IBDesignable reviewed (approved):
Спасибо за работу – зачет!)

После апрува можно вынести в отдельные файлы все дополнительные классы и структуры :) и затем вмерджить ветку.

Commit:
Правки по финальному ревью
- раскидал всё по файликам.
- вынес обработку следующей загадки в отдельную функцию
- использую в Alert weak self
- extension для GameState для упрощения кода
- избавился от -1 у currentRiddleIndex и добавил assert
- добавил имена переменным в enum
- переиспользую NumberFormatter
